internal/controllers/core/kubernetesdiscovery/reconciler.go

```diff
@@ -293,27 +308,29 @@ func (w *Reconciler) cleanupAbandonedNamespaces() {
 // This ensures it can be safely called by reconcile on each invocation for any namespace that the watcher cares about.
 // Additionally, for efficiency, duplicative watches on the same namespace will not be created; see watchedNamespaces
 // for more details.
-func (w *Reconciler) setupNamespaceWatch(ctx context.Context, ns k8s.Namespace, watcherKey watcherID) {
-	if watcher, ok := w.watchedNamespaces[ns]; ok {
+func (w *Reconciler) setupNamespaceWatch(ctx context.Context, nsKey clusterNamespace, watcherKey watcherID, kCli k8s.Client) {
+	if watcher, ok := w.watchedNamespaces[nsKey]; ok {
 		// already watching this namespace -- just add this watcher to the list for cleanup tracking
 		watcher.watchers[watcherKey] = true
 		return
 	}
 
-	ch, err := w.kCli.WatchPods(ctx, ns)
+	ns := nsKey.namespace
+	ch, err := kCli.WatchPods(ctx, ns)
 	if err != nil {
 		err = errors.Wrapf(err, "Error watching pods. Are you connected to kubernetes?\nTry running `kubectl get pods -n %q`", ns)
 		w.dispatcher.Dispatch(store.NewErrorAction(err))
 		return
 	}
 
 	ctx, cancel := context.WithCancel(ctx)
-	w.watchedNamespaces[ns] = nsWatch{
+	w.watchedNamespaces[nsKey] = nsWatch{
 		watchers: map[watcherID]bool{watcherKey: true},
 		cancel:   cancel,
 	}
 
-	go w.dispatchPodChangesLoop(ctx, ch)
+	ownerFetcher := k8s.ProvideOwnerFetcher(ctx, kCli)
```

Comment:
I need to refactor this - we really need to share this with the event + service watchers so we don't do multiple owner lookups for everything.

My current thinking is is to add `OwnerFetcher OwnerFetcher()` to the `K8sClient` interface, but I'm open to suggestions.

Reply:
i think the reconciler is going to need some per-cluster state, so that the state can be reset if the cluster resets. maybe put this there?

Reply:
For clarity, within the KD reconciler, this will only get re-created if we stop and then later restart the watch for a namespace.

The complication at the moment is that we use the `OwnerFetcher` elsewhere and want to share the same instance [for a given `K8sClient`] to make use of its internal caching/dedupe logic.

I moved ownership of the `OwnerFetcher` to the `K8sClient` interface in #5310 to do that. If we want to keep them separate, the alternative is to add it to the client cache interface afaict. I'm not against that, it just seems a bit less ergonomic for usage (tho perhaps cleaner from a separation of concerns perspective)

Reply:
ya, i think there are two separate questions:
- sharing state from the same k8s cluster
- clearing state when the k8s cluster changes

For cases where you want both (1) and (2), i think it's reasonable to move the state into the k8s client. but i'd worry if we started moving non-shared state into the k8s client.